### PR TITLE
[webgpu] Fix fusedConv2d bias weight and add teset case

### PR DESCRIPTION
**Pull request by axinging**
Fix: https://github.com/tensorflow/tfjs/issues/3095
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/3096)

### CODE REVIEW

**Comment by axinging** [comment]
@annxingyuan PTAL.
It seems this passed on WebGL/CPU/WebGPU, but failed on wasm.

```
Failures:
1) fused conv2d test-wasm {} relu bias stride 2 x=[1,8,8,16] f=[3,3,16,1] s=[2,2] d=4 p=same
  Message:
    Error: Arrays differ: actual[18] = 19469980, expected[18] = 19469982.
    Actual:   8772361,8794324,8816282,8838243,10094857,10121428,10147994,10174563,11417353,11448532,11479706,11510883,7493041,7516132,7539218,7562307,19352324,19411156,19469980,19528804,20674820,20738260,20801696,20865124,21997316,22065364,22133408,22201444,13759921,13807588,13855250,13902915,29932280,30027988,30123688,30219364,31254776,31355092,31455400,31555684,32577272,32682196,32787116,32892004,20026800,20099044,20171284,20243524,16368305,16450532,16532754,16614979,16955056,17040356,17125650,17210948,17541808,17630180,17718544,17806916,10026273,10086724,10147170,10207619.
    Expected: 8772361,8794324,8816282,8838243,10094857,10121428,10147994,10174563,11417353,11448532,11479706,11510883,7493041,7516132,7539218,7562307,19352324,19411156,19469982,19528804,20674820,20738260,20801694,20865124,21997312,22065364,22133410,22201444,13759921,13807588,13855250,13902915,29932280,30027988,30123690,30219364,31254776,31355092,31455402,31555684,32577272,32682196,32787114,32892004,20026796,20099044,20171286,20243524,16368305,16450532,16532754,16614979,16955056,17040356,17125650,17210948,17541808,17630180,17718546,17806916,10026273,10086724,10147170,10207619.
  Stack:
    Error: Arrays differ: actual[18] = 19469980, expected[18] = 19469982.
        at <Jasmine>
        at expectArraysPredicate (/workspace/tfjs-core/src/test_util.ts:90:13)
        at expectArraysClose (/workspace/tfjs-core/src/test_util.ts:32:10)
        at Object.<anonymous> (/workspace/tfjs-core/src/ops/fused_test.ts:730:8)
        at step (/workspace/tfjs-core/dist/ops/fused_test.js:48:23)
        at Object.next (/workspace/tfjs-core/dist/ops/fused_test.js:29:53)
        at fulfilled (/workspace/tfjs-core/dist/ops/fused_test.js:20:58)

2) fused conv2d test-wasm {} relu bias stride 2 x=[1,8,8,16] f=[3,3,16,1] s=[2,2] d=8 p=same
  Message:
    Error: Arrays differ: actual[2] = 17566680, expected[2] = 17566682.
    Actual:   17522760,17544724,17566680,17588644,17610608,17632566,17654524,17676488,20163140,20189716,20216284,20242852,20269420,20295990,20322560,20349128,22803524,22834708,22865888,22897060,22928236,22959414,22990596,23021768,14962993,14986084,15009170,15032259,15055353,15078438,15101525,15124616,38645816,38704656,38763500,38822308,38881116,38939968,38998800,39057608,41286200,41349640,41413100,41476516,41539932,41603392,41666832,41730248,43926584,43994632,44062700,44130724,44198748,44266816,44334864,44402888,27472168,27519844,27567512,27615172,27662832,27710502,27758172,27805832,59768880,59864568,59960308,60055972,60151636,60247384,60343064,60438728,62409264,62509560,62609908,62710180,62810452,62910808,63011096,63111368,65049640,65154544,65259508,65364388,65469268,65574232,65679128,65784008,39981352,40053600,40125852,40198084,40270316,40342568,40414816,40487048,32654372,32736612,32818844,32901060,32983276,33065510,33147744,33229960,33824804,33910116,33995424,34080708,34165996,34251304,34336608,34421896,34995236,35083624,35171996,35260356,35348716,35437096,35525472,35613832,19992096,20052548,20112996,20173444,20233896,20294342,20354792,20415240.
    Expected: 17522760,17544724,17566682,17588644,17610608,17632566,17654524,17676488,20163140,20189716,20216286,20242852,20269420,20295990,20322560,20349128,22803520,22834708,22865890,22897060,22928232,22959414,22990596,23021768,14962993,14986084,15009170,15032259,15055353,15078438,15101525,15124616,38645816,38704652,38763504,38822308,38881116,38939968,38998796,39057608,41286200,41349644,41413104,41476516,41539932,41603392,41666828,41730248,43926584,43994628,44062704,44130724,44198748,44266824,44334868,44402888,27472168,27519844,27567514,27615172,27662832,27710502,27758172,27805832,59768880,59864564,59960312,60055972,60151636,60247384,60343060,60438728,62409256,62509556,62609912,62710180,62810452,62910808,63011100,63111368,65049640,65154548,65259512,65364388,65469268,65574232,65679132,65784008,39981352,40053596,40125856,40198084,40270316,40342576,40414812,40487048,32654372,32736612,32818846,32901060,32983276,33065510,33147744,33229960,33824804,33910116,33995424,34080708,34165996,34251304,34336608,34421896,34995236,35083620,35172000,35260356,35348716,35437096,35525468,35613832,19992096,20052548,20112994,20173444,20233896,20294342,20354788,20415240.
  Stack:
    Error: Arrays differ: actual[2] = 17566680, expected[2] = 17566682.
        at <Jasmine>
        at expectArraysPredicate (/workspace/tfjs-core/src/test_util.ts:90:13)
        at expectArraysClose (/workspace/tfjs-core/src/test_util.ts:32:10)
        at Object.<anonymous> (/workspace/tfjs-core/src/ops/fused_test.ts:826:8)
        at step (/workspace/tfjs-core/dist/ops/fused_test.js:48:23)
        at Object.next (/workspace/tfjs-core/dist/ops/fused_test.js:29:53)
        at fulfilled (/workspace/tfjs-core/dist/ops/fused_test.js:20:58)

3) fused conv2d test-wasm {} relu bias stride 2 x=[1,8,8,16] f=[3,3,16,1] s=[2,2] d=32 p=same
  Message:
    Error: Arrays differ: actual[1] = 70047104, expected[1] = 70047112.
    Actual:   70025136,70047104,70069104,70091040,70112984,70134984,70156944,70178888,70200824,70222776,70244792,70266720,70288656,70310672,70332624,70354560,70376504,70398464,70420472,70442416,70464336,70486344,70508296,70530240,70552176,70574136,70596144,70618080,70640008,70662016,70683976,70705920,80572848,80599408,80626032,80652560,80679120,80705752,80732304,80758856,80785392,80811960,80838576,80865136,80891664,80918288,80944848,80971392,80997952,81024496,81051120,81077664,81104208,81130840,81157392,81183936,81210480,81237032,81263664,81290224,81316752,81343376,81369936,81396480,91120560,91151728,91182960,91214096,91245264,91276504,91307664,91338824,91369968,91401144,91432384,91463552,91494672,91525904,91557072,91588224,91619376,91650544,91681792,91712912,91744080,91775320,91806480,91837632,91868784,91899944,91931184,91962352,91993488,92024720,92055888,92087040,59782688,59805784,59828892,59851972,59875052,59898160,59921248,59944328,59967400,59990488,60013604,60036680,60059748,60082864,60105952,60129028,60152104,60175192,60198308,60221384,60244452,60267576,60290648,60313728,60336800,60359888,60383004,60406076,60429156,60452272,60475352,60498428,154406848,154465632,154524544,154583264,154642128,154701024,154759840,154818592,154877424,154936224,154995136,155054000,155112704,155171616,155230400,155289216,155348048,155406816,155465728,155524464,155583296,155642240,155701008,155759840,155818592,155877408,155936320,155995168,156053888,156112800,156171568,156230400,164954560,165017952,165081456,165144800,165208272,165271808,165335200,165398560,165461984,165525408,165588912,165652400,165715728,165779232,165842624,165906048,165969472,166032864,166096368,166159696,166223184,166286720,166350112,166413536,166476896,166540320,166603824,166667296,166730640,166794144,166857536,166920960,175502240,175570272,175638384,175706336,175774416,175842560,175910560,175978528,176046560,176114592,176182704,176250800,176318736,176386848,176454880,176522880,176590912,176658912,176727024,176794960,176863056,176931200,176999200,177067264,177135200,177203200,177271344,177339424,177407376,177475488,177543520,177611520,109745696,109793360,109841056,109888672,109936352,109984056,110031712,110079368,110127008,110174664,110222384,110270048,110317664,110365376,110413024,110460672,110508320,110555968,110603696,110651312,110698976,110746696,110794336,110841984,110889632,110937272,110984992,111032672,111080272,111128000,111175648,111223296,238788512,238884192,238979952,239075520,239171280,239267072,239362720,239458240,239554016,239649664,239745488,239841232,239936784,240032576,240128224,240223872,240319520,240415168,240510992,240606512,240702288,240798080,240893728,240989504,241085024,241180672,241276464,241372224,241467760,241563584,241659232,241754880,249336224,249436480,249536912,249637056,249737424,249837824,249938112,250038208,250138592,250238848,250339280,250439632,250539792,250640192,250740448,250840704,250940960,251041216,251141648,251241776,251342160,251442560,251542816,251643200,251743296,251843584,251943984,252044352,252144496,252244928,252345184,252445440,259883936,259988832,260093840,260198560,260303568,260408576,260513472,260618176,260723168,260828032,260933072,261038032,261142800,261247776,261352672,261457536,261562400,261667296,261772304,261877040,261982032,262087040,262191904,262296896,262401600,262506496,262611504,262716512,262821232,262926272,263031104,263136000,159708736,159780928,159853232,159925408,159997664,160069952,160142192,160214400,160286624,160358848,160431152,160503408,160575584,160647872,160720064,160792320,160864576,160936768,161009072,161081232,161153504,161225792,161298016,161370272,161442448,161514688,161586992,161659232,161731408,161803712,161875904,161948160,130370848,130453056,130535344,130617488,130699744,130782024,130864240,130946440,131028640,131110840,131193136,131275376,131357520,131439824,131522032,131604224,131686432,131768640,131850928,131933088,132015328,132097608,132179808,132262016,132344208,132426424,132508720,132590960,132673104,132755392,132837600,132919808,135043360,135128640,135214000,135299216,135384544,135469888,135555184,135640448,135725728,135811008,135896368,135981696,136066912,136152256,136237536,136322816,136408096,136493376,136578736,136663952,136749280,136834624,136919904,137005184,137090464,137175744,137261088,137346416,137431648,137516992,137602272,137687552,139715872,139804224,139892656,139980944,140069344,140157776,140246112,140334464,140422800,140511168,140599600,140687984,140776272,140864704,140953056,141041408,141129760,141218112,141306544,141394832,141483232,141571680,141660016,141748352,141836688,141925040,142013488,142101856,142190160,142278592,142366944,142455296,79787040,79847472,79907952,79968384,80028832,80089304,80149744,80210184,80270624,80331064,80391536,80451984,80512400,80572880,80633328,80693760,80754208,80814640,80875120,80935552,80996000,81056472,81116896,81177344,81237776,81298216,81358704,81419136,81479568,81540048,81600480,81660928.
    Expected: 70025136,70047112,70069104,70091040,70112984,70134984,70156952,70178888,70200824,70222784,70244792,70266728,70288656,70310656,70332624,70354560,70376504,70398464,70420472,70442408,70464336,70486352,70508296,70530240,70552168,70574136,70596144,70618080,70640016,70662032,70683976,70705920,80572848,80599416,80626032,80652560,80679128,80705736,80732312,80758856,80785400,80811968,80838584,80865144,80891664,80918288,80944848,80971392,80997944,81024496,81051128,81077656,81104208,81130832,81157384,81183936,81210472,81237048,81263664,81290224,81316752,81343376,81369928,81396480,91120560,91151736,91182960,91214080,91245272,91276504,91307672,91338824,91369976,91401136,91432376,91463560,91494672,91525904,91557072,91588224,91619384,91650544,91681784,91712904,91744080,91775328,91806472,91837632,91868776,91899944,91931184,91962368,91993488,92024720,92055880,92087040,59782688,59805780,59828892,59851972,59875052,59898164,59921252,59944328,59967396,59990488,60013600,60036680,60059752,60082864,60105952,60129028,60152104,60175188,60198304,60221384,60244456,60267576,60290652,60313724,60336796,60359888,60383004,60406076,60429152,60452272,60475356,60498428,154406848,154465632,154524544,154583264,154642128,154701024,154759840,154818592,154877408,154936224,154995136,155054000,155112704,155171616,155230400,155289216,155348032,155406816,155465728,155524448,155583296,155642208,155701024,155759840,155818592,155877408,155936320,155995168,156053888,156112800,156171584,156230400,164954560,165017952,165081456,165144800,165208288,165271776,165335200,165398560,165461984,165525408,165588912,165652400,165715728,165779232,165842624,165906048,165969472,166032864,166096368,166159712,166223184,166286688,166350112,166413536,166476896,166540320,166603824,166667296,166730640,166794144,166857536,166920960,175502240,175570272,175638384,175706336,175774432,175842560,175910560,175978496,176046560,176114560,176182704,176250800,176318736,176386848,176454880,176522880,176590880,176658912,176727024,176794976,176863056,176931200,176999200,177067264,177135200,177203200,177271344,177339456,177407376,177475488,177543520,177611520,109745696,109793352,109841056,109888672,109936360,109984056,110031720,110079368,110127016,110174672,110222376,110270056,110317664,110365376,110413024,110460672,110508328,110555968,110603688,110651304,110698976,110746688,110794328,110841984,110889624,110937288,110984992,111032672,111080288,111128000,111175640,111223296,238788512,238884160,238979952,239075520,239171296,239267072,239362720,239458240,239554016,239649664,239745456,239841232,239936784,240032576,240128224,240223872,240319520,240415168,240510960,240606528,240702288,240798080,240893728,240989504,241085024,241180672,241276464,241372224,241467792,241563584,241659232,241754880,249336224,249436480,249536880,249637024,249737440,249837824,249938080,250038208,250138592,250238848,250339248,250439632,250539792,250640192,250740448,250840704,250940960,251041216,251141616,251241792,251342160,251442560,251542816,251643200,251743328,251843584,251943984,252044384,252144528,252244928,252345184,252445440,259883936,259988832,260093808,260198560,260303584,260408576,260513440,260618176,260723168,260828032,260933040,261038032,261142800,261247776,261352672,261457536,261562400,261667296,261772272,261877056,261982032,262087040,262191904,262296896,262401632,262506496,262611504,262716512,262821264,262926272,263031104,263136000,159708736,159780928,159853232,159925408,159997664,160069952,160142176,160214368,160286624,160358848,160431152,160503408,160575568,160647872,160720064,160792320,160864576,160936768,161009072,161081248,161153488,161225792,161298016,161370272,161442464,161514688,161586992,161659232,161731408,161803712,161875904,161948160,130370848,130453064,130535344,130617488,130699736,130782024,130864232,130946440,131028648,131110848,131193144,131275384,131357520,131439808,131522016,131604224,131686440,131768640,131850936,131933080,132015312,132097616,132179800,132262016,132344216,132426424,132508720,132590960,132673104,132755392,132837592,132919808,135043360,135128640,135214000,135299216,135384544,135469888,135555168,135640448,135725728,135811008,135896352,135981680,136066912,136152256,136237536,136322816,136408096,136493376,136578720,136663968,136749280,136834624,136919904,137005184,137090464,137175744,137261088,137346416,137431648,137516992,137602272,137687552,139715872,139804224,139892656,139980944,140069344,140157776,140246112,140334464,140422800,140511152,140599600,140687984,140776272,140864704,140953056,141041408,141129760,141218112,141306544,141394848,141483216,141571680,141660016,141748352,141836688,141925040,142013488,142101856,142190160,142278608,142366944,142455296,79787040,79847480,79907952,79968384,80028824,80089288,80149736,80210184,80270632,80331072,80391544,80451976,80512400,80572880,80633312,80693760,80754216,80814640,80875128,80935560,80995984,81056464,81116888,81177344,81237784,81298232,81358704,81419136,81479568,81540048,81600472,81660928.
  Stack:
    Error: Arrays differ: actual[1] = 70047104, expected[1] = 70047112.
        at <Jasmine>
        at expectArraysPredicate (/workspace/tfjs-core/src/test_util.ts:90:13)
        at expectArraysClose (/workspace/tfjs-core/src/test_util.ts:32:10)
        at Object.<anonymous> (/workspace/tfjs-core/src/ops/fused_test.ts:1005:8)
        at step (/workspace/tfjs-core/dist/ops/fused_test.js:48:23)
        at Object.next (/workspace/tfjs-core/dist/ops/fused_test.js:29:53)
        at fulfilled (/workspace/tfjs-core/dist/ops/fused_test.js:20:58)

4) fused conv2d test-wasm {} prelu bias stride 2 x=[1,8,8,16] f=[3,3,16,1] s=[2,2] d=8 p=same
  Message:
    Error: Arrays differ: actual[6] = -175821408, expected[6] = -175821440.
    Actual:   -24805912,-49715608,-74729104,-99846512,-125067760,-150392704,-175821408,-201354176,-22165524,-44425624,-66780312,-89229680,-111773664,-134412160,-157145216,-179973056,-19525136,-39135640,-58831536,-78612848,-98479560,-118431616,-138469040,-158591936,-10104239,-20260792,-30469674,-40730868,-51044356,-61410204,-71828368,-82298816,-3682823,-7395736,-11138754,-14911860,-18715036,-22548348,-26411748,-30305216,-1042439,-2105752,-3189954,-4295028,-5420955,-6567804,-7735525,-8924096,1597945,1592116,1586282,1580451,1574625,1568790,1562957,1557128,2404945,2403364,2401778,2400195,2398617,2397030,2395445,2393864,17440248,17462068,17483884,17505700,17527520,17549334,17571148,17592968,20080628,20107060,20133484,20159908,20186332,20212758,20239184,20265608,22721008,22752052,22783088,22814116,22845144,22876182,22907220,22938248,14914129,14937124,14960114,14983107,15006105,15029094,15052085,15075080,13890897,13923876,13956850,13989827,14022809,14055782,14088757,14121736,15061329,15097380,15133426,15169475,15205529,15241574,15277621,15313672,16231761,16270884,16310002,16349123,16388249,16427366,16466485,16505608,9584353,9611972,9639586,9667203,9694825,9722438,9750053,9777672.
    Expected: -24805912,-49715608,-74729104,-99846512,-125067760,-150392704,-175821440,-201354176,-22165524,-44425624,-66780320,-89229680,-111773664,-134412160,-157145248,-179973056,-19525136,-39135640,-58831528,-78612848,-98479560,-118431616,-138469040,-158591936,-10104239,-20260792,-30469674,-40730868,-51044356,-61410204,-71828368,-82298816,-3682823,-7395736,-11138754,-14911860,-18715036,-22548348,-26411748,-30305216,-1042439,-2105752,-3189954,-4295028,-5420955,-6567804,-7735525,-8924096,1597945,1592116,1586282,1580451,1574625,1568790,1562957,1557128,2404945,2403364,2401778,2400195,2398617,2397030,2395445,2393864,17440248,17462068,17483882,17505700,17527520,17549334,17571148,17592968,20080628,20107060,20133486,20159908,20186332,20212758,20239184,20265608,22721008,22752052,22783090,22814116,22845144,22876182,22907220,22938248,14914129,14937124,14960114,14983107,15006105,15029094,15052085,15075080,13890897,13923876,13956850,13989827,14022809,14055782,14088757,14121736,15061329,15097380,15133426,15169475,15205529,15241574,15277621,15313672,16231761,16270884,16310002,16349123,16388249,16427366,16466485,16505608,9584353,9611972,9639586,9667203,9694825,9722438,9750053,9777672.
  Stack:
    Error: Arrays differ: actual[6] = -175821408, expected[6] = -175821440.
        at <Jasmine>
        at expectArraysPredicate (/workspace/tfjs-core/src/test_util.ts:90:13)
        at expectArraysClose (/workspace/tfjs-core/src/test_util.ts:32:10)
        at Object.<anonymous> (/workspace/tfjs-core/src/ops/fused_test.ts:1189:8)
        at step (/workspace/tfjs-core/dist/ops/fused_test.js:48:23)
        at Object.next (/workspace/tfjs-core/dist/ops/fused_test.js:29:53)
        at fulfilled (/workspace/tfjs-core/dist/ops/fused_test.js:20:58)
```

**Comment by annxingyuan** [comment]
Hi @axinging - I have not tried handpose in WASM. I assumed it wouldn't work because we haven't implemented the rotation kernel in WASM, but I didn't realize there was also a correctness issue with conv2d in WASM... thank you for filing that bug. 

Are you saying that it does not work to exclude tests in the WASM backend? Do you mean that even though you include the name of the test in setup_test, the test still runs?

File: tfjs-core/src/ops/fused_test.ts [hunk 1]
```diff
@@ -640,6 +655,568 @@ describeWithFlags('fused conv2d', ALL_ENVS, () => {
...
+      filter: w,
+      strides: stride,
+      pad,
+      dataFormat: 'NHWC',
+      dilations: [1, 1],
+      activation: 'relu'
+    });
+    expect(result.shape).toEqual([1, 4, 4, 4]);
+    expectArraysClose(
+        await result.data(), new Float32Array([
```
Comment (annxingyuan) [inline tfjs-core/src/ops/fused_test.ts:686]: Would it be possible to reduce the number of test cases to 2? 

Also, these tests will probably fail in Safari due to overflow issues in WebGL (see: https://github.com/tensorflow/tfjs/issues/3143). While we work on fixing the underlying issue, would you mind changing your input generator so that the outputs stay under ~1000?

**Comment by annxingyuan** [comment]
@axinging Thanks for catching this failure! However I'm a little bit confused - how does `getPreluActivationWeightsAtOutCoords` fail? I would rather we fixed the `getPreluActivationWeightsAtOutCoords` function to work without additional arguments - the point of that helper function is that it figures out where to sample from without additional arguments. Currently, the source code for the activation function in this shader matches WebGL. I would like for us to maintain that if possible.

**Comment by axinging** [comment]
> @axinging Thanks for catching this failure! However I'm a little bit confused - how does `getPreluActivationWeightsAtOutCoords` fail? I would rather we fixed the `getPreluActivationWeightsAtOutCoords` function to work without additional arguments - the point of that helper function is that it figures out where to sample from without additional arguments. Currently, the source code for the activation function in this shader matches WebGL. I would like for us to maintain that if possible.

If I understands correctly, the bias and PreluActivationWeights is applied to every output by depth.
So it should be addressed by the depth of output.
When workPerThread is 1, below getBiasAtOutCoords (The same as getPreluActivationWeightsAtOutCoords) works.  In this case, the depth of output equals coords[3.
```
    float getBiasAtOutCoords() {
      ivec4 coords = getOutputCoords();

      return float(bias[getFlatIndex(int(coords[3]), biasShape)]);
    }
```

However, when workPerThread is >1, coords[3] no more equals output depth. The real depth of output is from col of mm_write:
```
        void mm_write(int row, int col, float value) {
          ivec4 outCoord = ivec4(
              batch,
              row / outShape[2],
              row % outShape[2],
              col);
          ${addBiasSnippet}
          ${applyActivationSnippet}
          result[getFlatIndex(outCoord, outShape)] = value;
        }
```

**Comment by annxingyuan** [comment]
@axinging I see - can we pass in the entire `outCoord`, instead of `ivec4(0,0,0,col)`? Doesn't `ivec4(0,0,0,col)` assume that we're broadcasting the bias? Our `getVarAtOutputCoords` should be able to handle broadcasting on its own.

**Comment by axinging** [comment]
> outCoord
Yes, use outCoord is more readable. Updated, PTAL.
